arangod/Graph/GraphOperations.cpp

```diff
@@ -605,119 +590,218 @@ OperationResult GraphOperations::updateEdge(const std::string& definitionName,
+  std::vector<std::string> writeCollections;
+  std::vector<std::string> exclusiveCollections;
+
+  if (foundEdgeDefinition) {
+    readCollections.emplace_back(fromCollectionName);
+    readCollections.emplace_back(toCollectionName);
+  }
+  writeCollections.emplace_back(definitionName);
+
+  transaction::Options trxOptions;
+  trxOptions.waitForSync = waitForSync;
+
+  auto trx = std::make_unique<transaction::Methods>(ctx(), readCollections, writeCollections,
+                                                    exclusiveCollections, trxOptions);
+
+  Result tRes = trx->begin();
+
+  if (!tRes.ok()) {
+    return std::make_pair(OperationResult(tRes), nullptr);
+  }
+
+  if (foundEdgeDefinition) {
+    res = validateEdgeVertices(fromCollectionName, fromCollectionKey,
+                               toCollectionName, toCollectionKey, *trx.get());
+
+    if (res.fail()) {
+      return std::make_pair(std::move(res), nullptr);
+    }
+  }
+
+  return std::make_pair(std::move(res), std::move(trx));
+}
+
+OperationResult GraphOperations::validateEdgeVertices(
+    const std::string& fromCollectionName, const std::string& fromCollectionKey,
+    const std::string& toCollectionName, const std::string& toCollectionKey,
+    transaction::Methods& trx) {
+  VPackBuilder bT;
+  {
+    VPackObjectBuilder guard(&bT);
+    bT.add(StaticStrings::KeyString, VPackValue(toCollectionKey));
+  }
+
+  VPackBuilder bF;
+  {
+    VPackObjectBuilder guard(&bF);
+    bF.add(StaticStrings::KeyString, VPackValue(fromCollectionKey));
+  }
+
+  OperationOptions options;
+  OperationResult resultFrom = trx.document(fromCollectionName, bF.slice(), options);
+  OperationResult resultTo = trx.document(toCollectionName, bT.slice(), options);
+
+  // actual result doesn't matter here
+  if (!resultFrom.ok()) {
+    trx.finish(resultFrom.result);
+    return OperationResult(TRI_ERROR_ARANGO_DOCUMENT_NOT_FOUND);
+  } else if (!resultTo.ok()) {
+    trx.finish(resultTo.result);
+    return OperationResult(TRI_ERROR_ARANGO_DOCUMENT_NOT_FOUND);
+  }
+  return OperationResult(TRI_ERROR_NO_ERROR);
+}
+
+std::pair<OperationResult, bool> GraphOperations::validateEdgeContent(
+    const VPackSlice& document, std::string& fromCollectionName, std::string& fromCollectionKey,
+    std::string& toCollectionName, std::string& toCollectionKey, bool isUpdate) {
   VPackSlice fromStringSlice = document.get(StaticStrings::FromString);
   VPackSlice toStringSlice = document.get(StaticStrings::ToString);
 
   if (fromStringSlice.isNone() || toStringSlice.isNone()) {
-    return OperationResult(TRI_ERROR_ARANGO_INVALID_EDGE_ATTRIBUTE);
+    if (isUpdate) {
+      return std::make_pair(OperationResult(TRI_ERROR_NO_ERROR), false);
+    }
+    return std::make_pair(OperationResult(TRI_ERROR_ARANGO_INVALID_EDGE_ATTRIBUTE), false);
   }
   std::string fromString = fromStringSlice.copyString();
   std::string toString = toStringSlice.copyString();
 
   size_t pos = fromString.find('/');
```

Comment:
Is there a centrally provided function that validates document identifiers? This ad-hoc parsing of names looks a bit fragile to me.